File: frontend/src/api/behaviorTemplates.ts
import { api } from "./client";
import type { Workflow } from "../types/campaign";
import type { BehaviorTemplate, BehaviorTemplateInput } from "../types/behaviorTemplate";

export async function listBehaviorTemplates(): Promise<BehaviorTemplate[]> {
  const response = await api.get<BehaviorTemplate[]>("/behavior-templates");
  return response.data;
}

export async function getBehaviorTemplate(templateId: string): Promise<BehaviorTemplate> {
  const response = await api.get<BehaviorTemplate>(`/behavior-templates/${templateId}`);
  return response.data;
}

export async function createBehaviorTemplate(input: BehaviorTemplateInput): Promise<BehaviorTemplate> {
  const response = await api.post<BehaviorTemplate>("/behavior-templates", input);
  return response.data;
}

export async function deleteBehaviorTemplate(templateId: string): Promise<void> {
  await api.delete(`/behavior-templates/${templateId}`);
}

export async function applyBehaviorTemplate(campaignId: string, templateId: string): Promise<Workflow> {
  const response = await api.post<Workflow>(`/campaigns/${campaignId}/apply-template`, {
    template_id: templateId
  });
  return response.data;
}

File: backend/app/api/v1/behavior_templates.py
from uuid import UUID

from fastapi import APIRouter, Depends, Response, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.session import get_session
from app.schemas.behavior_template import (
    ApplyBehaviorTemplateRequest,
    BehaviorTemplateCreate,
    BehaviorTemplateRead,
)
from app.schemas.workflow import WorkflowRead
from app.services.behavior_template_service import BehaviorTemplateService

router = APIRouter(tags=["behavior-templates"])


def service(session: AsyncSession = Depends(get_session)) -> BehaviorTemplateService:
    return BehaviorTemplateService(session)


@router.get("/behavior-templates", response_model=list[BehaviorTemplateRead])
async def list_behavior_templates(
    template_service: BehaviorTemplateService = Depends(service),
) -> list[BehaviorTemplateRead]:
    return await template_service.list_templates()


@router.get("/behavior-templates/{template_id}", response_model=BehaviorTemplateRead)
async def get_behavior_template(
    template_id: UUID,
    template_service: BehaviorTemplateService = Depends(service),
) -> BehaviorTemplateRead:
    return await template_service.get_template(template_id)


@router.post("/behavior-templates", response_model=BehaviorTemplateRead, status_code=status.HTTP_201_CREATED)
async def create_behavior_template(
    payload: BehaviorTemplateCreate,
    template_service: BehaviorTemplateService = Depends(service),
) -> BehaviorTemplateRead:
    return await template_service.create_template(payload)


@router.delete("/behavior-templates/{template_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_behavior_template(
    template_id: UUID,
    template_service: BehaviorTemplateService = Depends(service),
) -> Response:
    await template_service.delete_template(template_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.post("/campaigns/{campaign_id}/apply-template", response_model=WorkflowRead)
async def apply_behavior_template(
    campaign_id: UUID,
    payload: ApplyBehaviorTemplateRequest,
    template_service: BehaviorTemplateService = Depends(service),
) -> WorkflowRead:
    return await template_service.apply_template(campaign_id, payload.template_id)
